lib/aliquot-pay.rb

```diff
@@ -72,36 +72,63 @@ def self.four_byte_length(str)
     [str.length].pack('V')
   end
 
+  def self.generate_signature(*args)
+    args.map do |s|
+      four_byte_length(s) + s
+    end.join('')
+  end
+
   def self.signature_string(
     message,
-    recipient_id     = DEFAULTS[:merchant_id],
-    sender_id        = DEFAULTS[:info],
-    protocol_version = 'ECv1'
+    recipient_id:     "merchant:#{DEFAULTS[:merchant_id]}",
+    sender_id:        DEFAULTS[:info],
+    protocol_version: 'ECv1'
   )
 
-    four_byte_length(sender_id) +
-      sender_id +
-      four_byte_length(recipient_id) +
-      recipient_id +
-      four_byte_length(protocol_version) +
-      protocol_version +
-      four_byte_length(message) +
-      message
+    generate_signature(sender_id, recipient_id, protocol_version, message)
   end
 
-  # payment:: Google Pay token as a ruby Hash
-  # signing_key::           OpenSSL::PKEY::EC
-  # recipient::             OpenSSL::PKey::EC
-  # encrypted_message::     SignedMessage
-  def self.generate_token(payment, signing_key, recipient, signed_message = nil)
-    signed_message ||= JSON.unparse(AliquotPay.encrypt(JSON.unparse(payment), recipient))
-
-    signature_string = AliquotPay.signature_string(signed_message)
+  # payment::        Google Pay token as a ruby Hash
+  # signing_key::    OpenSSL::PKEY::EC
+  # recipient::      OpenSSL::PKey::EC
+  # signed_message:: Pass a customized message to sign as signed messaged.
+  def self.generate_token_ecv1(payment, signing_key, recipient, signed_message = nil)
+    cipher = OpenSSL::Cipher::AES128.new(:CTR)
+    signed_message ||= JSON.unparse(encrypt(JSON.unparse(payment), recipient, cipher))
+    signature_string = signature_string(signed_message)
 
     {
       'protocolVersion' => 'ECv1',
-      'signature' =>       AliquotPay.sign(signing_key, signature_string),
+      'signature' =>       sign(signing_key, signature_string),
       'signedMessage' =>   signed_message,
     }
   end
+
+  def self.generate_token_ecv2(payment, signing_key, intermediate_key, recipient,
+                               signed_message: nil, expire_time:  "#{Time.now.to_i + 3600}000")
+    cipher = OpenSSL::Cipher::AES256.new(:CTR)
+    signed_message ||= JSON.unparse(encrypt(JSON.unparse(payment), recipient, cipher))
```

Comment:
Can you explain the two instances of JSON unparsing here? I am a bit confused :thinking:

Reply:
The inner `payment` is `#to_json`'ed (`JSON.unparse`'d) before it's encrypted ...

Reply:
Exactly. The `payment` is added as serialized JSON, this is done because it is what is verified by `tag`, a MAC.
This results in a signed message, which is also a serialized JSON string in the final token, since it is signed by Google.